src/clj/game/cards/upgrades.clj

```diff
@@ -653,6 +653,31 @@
       :events {:pre-steal-cost ohg
                :post-access-card {:effect (effect (clear-persistent-flag! target :can-steal))}}})
 
+   "Overseer Matrix"
+   (let [om {:req (req (in-same-server? card target))
+             :delayed-completion true
+             :effect (effect (show-wait-prompt :runner "Corp to use Overseer Matrix")
+                             (continue-ability
+                               {:optional
+                                {:prompt "Pay 1 [Credits] to use Overseer Matrix ability?"
+                                 :player :corp
+                                 :end-effect (effect (clear-wait-prompt :runner)
+                                                     (effect-completed eid))
+                                 :yes-ability {:cost [:credit 1]
+                                               :msg "give the Runner 1 tag"
+                                               :delayed-completion true
+                                               :effect (req (tag-runner state :runner eid 1))}}}
+                               card nil))}]
+     {:trash-effect
+      {:req (req (and (= :servers (first (:previous-zone card))) (:run @state)))
+       :effect (effect (register-events {:runner-trash (assoc om :req (req (or (= (:zone (get-nested-host target)) (:previous-zone card))
```

Comment:
Why does `in-same-server?` not work here?

Reply:
This code fires after Overseer Matrix has been trashed, so I think we need to look where it was, not where it is (Archives).